### PR TITLE
Mcts updates

### PR DESCRIPTION
**Pull request by nashirj**
Lots of updates!!

### CODE REVIEW

File: chess-AI/graph.py [hunk 1]
```diff
@@ -0,0 +1,17 @@
+import matplotlib.pyplot as plt
```
Comment (nashirj) [inline chess-AI/graph.py:1]: This file could be called `visualization.py` to make its use more clear!

File: chess-AI/mcts.py [hunk 2]
```diff
@@ -1,8 +1,11 @@
 from math import sqrt
-
+# import pandas as pd
```
Comment (nashirj) [inline chess-AI/mcts.py:2]: Can remove if not using pandas

File: chess-AI/mcts.py [hunk 3]
```diff
@@ -1,8 +1,11 @@
 from math import sqrt
-
```
Comment (nashirj) [inline chess-AI/mcts.py:2]: These blank lines should be kept. Should have a blank line between each "type" of import; builtin includes at the top; 3rd party includes after one newline; project specific includes after one newline. See [style guide](https://google.github.io/styleguide/pyguide.html#313-imports-formatting)

File: chess-AI/mcts.py [hunk 4]
```diff
@@ -1,8 +1,11 @@
...
+# import pandas as pd
 import numpy as np
-
 from state_representation import get_cnn_input
+from output_representation import PlayNetworkPolicyConverter
+import torch
+import random
```
Comment (nashirj) [inline chess-AI/mcts.py:7]: Similar to above comment, [style guide](https://google.github.io/styleguide/pyguide.html#313-imports-formatting) says: "Within each grouping, imports should be sorted lexicographically, ignoring case, according to each module’s full package path (the path in from path import ...). Code may optionally place a blank line between import sections."

File: chess-AI/mcts.py [hunk 5]
```diff
@@ -12,147 +15,125 @@ class Mcts:
         n_values: Dictionary storing the n values
         q_values: Dictionary storing the q values
         exploration: The exploration rate for the tree search
-        p_values: Dictionary storing the search probabilites  
-    """
+        p_values: Dictionary storing the search probabilites"""
```
Comment (nashirj) [inline chess-AI/mcts.py:18]: Closing `"""` [should be on newline](https://google.github.io/styleguide/pyguide.html#383-functions-and-methods)

File: chess-AI/mcts.py [hunk 6]
```diff
@@ -12,147 +15,125 @@ class Mcts:
...
+
+        best_move_index = np.argmax(n_values)[0]
+        best_move = moves[best_move_index]
+
+        return best_move
+
+    def set_current_qn_value(self, fen_string, uci_move):
+        """Returns q and n values if they exist, otherwise initializes them"""
+        if fen_string in self.state_values:
+            if uci_move in self.state_values[fen_string]:
```
Comment (nashirj) [inline chess-AI/mcts.py:46]: Can be condensed:
```python
if uci_move not in self.state_values[fen_string]:
    self.state_values[fen_string][uci_move] = [0, 0]
return self.state_values[fen_string][uci_move]
```

File: chess-AI/mcts.py [hunk 7]
```diff
@@ -12,147 +15,125 @@ class Mcts:
...
-        """
-        # Calculates the search probabilites based on the the number of times a node has
-        # been visited/sum of all nodes visited
-        
+
+    def update_qn(self, fen_string, uci_move, value):
+        """Update q and n values after an expansion"""
+        n_value, q_value = self.state_values[fen_string][uci_move]
+
+        if not isinstance(value, int):
```
Comment (nashirj) [inline chess-AI/mcts.py:81]: Could optionally have a comment explaining why you have this line

File: chess-AI/mcts.py [hunk 8]
```diff
@@ -12,147 +15,125 @@ class Mcts:
...
-            # Add the state to the list of visited nodes
             self.states_visited.append(fen_string)
 
-            # TODO: Get neural network state representation from fen_string
-            # input_state = get_cnn_input(fen_string)
-
-            # Return predictions and value from the nnet at the current state
-            value, policy = nnet.predict(board, None)
+            # Get predictions and value from the nnet at the current state
+            policy, value = nnet(get_cnn_input(board).float())
```
Comment (nashirj) [inline chess-AI/mcts.py:119]: Do we need this cast to float here? If it needs to be a float, I think that should be done within the definition of `get_cnn_input(board)` rather than here

File: chess-AI/mcts.py [hunk 9]
```diff
@@ -12,147 +15,125 @@ class Mcts:
...
+            policy = self.policy_converter.find_value_of_all_legal_moves(policy, board)
```
Comment (nashirj) [inline chess-AI/mcts.py:120]: Can we pls use a different variable name for this line and the line above? Maybe keep `policy` above and then `policy_legal_moves` here.

File: chess-AI/nn_layout.py [hunk 10]
```diff
@@ -23,116 +23,94 @@ class PlayNetwork(nn.Module):
...
+        self.res_blocks = [nn.Sequential(nn.Conv2d(in_channels=self.num_filters,
+                                                 out_channels=self.num_filters,
                                                  kernel_size=3,
                                                  padding=1,
                                                  bias=False), 
-                                       nn.BatchNorm2d(256), 
-                                       nn.ReLU(inplace=True), 
-                                       nn.Conv2d(in_channels=256,
-                                                 out_channels=256,
+                                       nn.BatchNorm2d(self.num_filters),
```
Comment (nashirj) [inline chess-AI/nn_layout.py:46]: Can we pls make this align with `nn.Conv2d` above (and same for subsequent lines)

Comment (nashirj) [reply chess-AI/nn_layout.py:46]: This whole block in general got slightly misaligned

File: chess-AI/nn_layout.py [hunk 11]
```diff
@@ -23,116 +23,94 @@ class PlayNetwork(nn.Module):
...
-        return value, move_values
-                
 def main():
-    model = PlayNetwork()
-    policy, value = model(torch.randn(1, 119, 8, 8))
-    policy = policy.reshape(8, 8, 73)
-    value = value.item()
-    print(policy)
-    print(value)
+    model = PlayNetwork().float()
```
Comment (nashirj) [inline chess-AI/nn_layout.py:104]: Why do we need `float()` here?

File: chess-AI/nn_layout.py [hunk 12]
```diff
@@ -23,116 +23,94 @@ class PlayNetwork(nn.Module):
...
 
     import chess
     board = chess.Board()
     policy_converter = PlayNetworkPolicyConverter()
 
+    policy, value = model(get_cnn_input(board).float())
```
Comment (nashirj) [inline chess-AI/nn_layout.py:110]: Same comment as above about if we need this to be `.float()` let's have that happen in `get_cnn_input()`

File: chess-AI/self_play_dataset.py [hunk 13]
```diff
@@ -0,0 +1,15 @@
...
+        self.mcts_probs = torch.tensor(mcts_probs)
+        self.mcts_evals = torch.tensor(mcts_evals)
+        self.boards = boards
+        self.num_moves = self.mcts_evals.size
+    
+    def __getitem__(self, index):
+        return self.mcts_probs[index], self.mcts_evals[index], self.boards[index]
+    
+    def __len__(self):
+        return self.num_moves
```
Comment (nashirj) [inline chess-AI/self_play_dataset.py:15]: Should have newline at end of file.

File: chess-AI/self_play_dataset.py [hunk 14]
```diff
@@ -0,0 +1,15 @@
+from torch.utils.data import Dataset
```
Comment (nashirj) [inline chess-AI/self_play_dataset.py:1]: If you're importing `torch`, just do `torch.utils.data.Dataset` below and get rid of this import.

File: chess-AI/state_representation.py [hunk 15]
```diff
@@ -1,6 +1,6 @@
 import torch
-import chess
 import numpy as np
+import chess
```
Comment (nashirj) [inline chess-AI/state_representation.py:3]: This should be
```python
import numpy as np
import torch

import chess
```
See above comment about import order

File: chess-AI/train.py [hunk 16]
```diff
@@ -1,10 +1,18 @@
 import random
 
 import chess
+import torch
+import numpy as np
+import time
+from torch.autograd import backward
+from torch.functional import Tensor
+from torch.utils.data import DataLoader, TensorDataset
```
Comment (nashirj) [inline chess-AI/train.py:9]: Same comment as above, either selectively import from torch for everything, or just import torch and then do `torch.*` as you need

File: chess-AI/train.py [hunk 17]
```diff
@@ -1,10 +1,18 @@
...
 
 from mcts import Mcts
 from nn_layout import PlayNetwork
 from output_representation import PlayNetworkPolicyConverter
+from state_representation import get_cnn_input
+# from self_play_dataset import SelfPlayDataset
```
Comment (nashirj) [inline chess-AI/train.py:15]: Can remove if not being used yet

File: chess-AI/train.py [hunk 18]
```diff
@@ -15,73 +23,144 @@ class MctsTrain:
...
-                # Need to assign the rewards to the examples
-                self.assign_rewards(board)
-                return
+                    # Perform mcts simulations
+                    for _ in range(self.mcts_simulations):
+                        self.mcts.search(board, nnet)
 
+                    # Gets the moves and policy from the mcts, as well as the individual move to take
+                    # NOTE: moves[i] corresponds to search_probs[i]
+                    moves, search_probs, move = self.mcts.find_search_probs(fen_string, 5)
```
Comment (nashirj) [inline chess-AI/train.py:66]: If you don't create a variable for the parameter corresponding to `5`, please include a parameter label in the call to `find_search_probs` i.e.`find_search_probs(fen_string, temperature=5)`

File: chess-AI/train.py [hunk 19]
```diff
@@ -15,73 +23,144 @@ class MctsTrain:
...
 
-            # Need to get the moves and policy from the mcts
-            # NOTE: moves[i] corresponds to search_probs[i]
-            moves, search_probs = self.mcts.find_search_probs(fen_string)
+    def training_episode(self, nnet, epochs):
+        with torch.no_grad():
+            # At each episode need to set the training example and network predictions to empty
+            self.mcts_probs = []
+            self.mcts_evals = []
+            self.boards = []
```
Comment (nashirj) [inline chess-AI/train.py:48]: Could add a comment describing that this is a 2d array where each 1d array consists of fen strings corresponding to one game

File: chess-AI/train.py [hunk 20]
```diff
@@ -15,73 +23,144 @@ class MctsTrain:
...
+            
+                    # Store board state (used in training loop)
+                    self.boards.append(fen_string)
+
+                    # Converts mcts search probabilites to (8,8,73) vector
+                    full_search_probs = self.policy_converter.compute_full_search_probs(moves, search_probs, board)
+                    
+                    # Adds entry to the training examples
+                    self.mcts_probs.append(full_search_probs)
+                    self.mcts_evals.append(None)
```
Comment (nashirj) [inline chess-AI/train.py:76]: Why append `None` here?

File: chess-AI/train.py [hunk 21]
```diff
@@ -15,73 +23,144 @@ class MctsTrain:
...
+
+                    # Makes the random action on the board, and gets fen string
+                    move = chess.Move.from_uci(move)
+                    board.push(move)
+                    print(board)
+
+                    # If the game is over, end the episode
+                    if board.is_game_over() or board.is_stalemate() or board.is_seventyfive_moves() or board.is_fivefold_repetition() or board.can_claim_draw():
```
Comment (nashirj) [inline chess-AI/train.py:84]: I'm not sure, but does `is_game_over()` handle all these cases?

File: chess-AI/train.py [hunk 22]
```diff
@@ -15,73 +23,144 @@ class MctsTrain:
...
+
+                self.mcts_probs = torch.tensor(np.array(self.mcts_probs)).float()
+                self.mcts_evals = torch.tensor(self.mcts_evals).float()
+
+                # Create iterable dataset with mcts labels
+                dataset = TensorDataset(self.mcts_probs, self.mcts_evals)
+                train_dl = DataLoader(dataset=dataset, batch_size=batch_size, shuffle=False)
+
+                # Define loss functions
+                loss_fn1 = torch.nn.CrossEntropyLoss()
```
Comment (nashirj) [inline chess-AI/train.py:96]: Pls change to `ce_loss_fn` and `mse_loss_fn` for readability

Comment (nashirj) [reply chess-AI/train.py:96]: You can then optionally make a list `loss_fns = [ce_loss_fn, mse_loss_fn]` and iterate over this list to compute loss for all the different loss functions you care about

File: chess-AI/train.py [hunk 23]
```diff
@@ -15,73 +23,144 @@ class MctsTrain:
...
     def assign_rewards(self, board):
         """Iterates through training examples and assigns rewards based on result of the game.
         """
-        for example in self.training_examples:
-            if board.outcome().winner:
-                example[2] = -1
-            else:
-                example[2] = 0
+        reward = 0
+        if not board.outcome() == None:
```
Comment (nashirj) [inline chess-AI/train.py:142]: `if board.outcome() is not None:`

File: chess-AI/train.py [hunk 24]
```diff
@@ -15,73 +23,144 @@ class MctsTrain:
...
+            if not board.outcome().winner == None:
```
Comment (nashirj) [inline chess-AI/train.py:143]: `if board.outcome().winner is not None:`

**Review by nashirj — COMMENTED** [review]
Overall this looks amazing!! Nice work everyone. Really clean code and a lot of good progress <3
